src/re2-stubs/__init__.pyi

```diff
@@ -117,79 +117,37 @@ def purge() -> None: ...
 
 class _Regexp(Generic[AnyStr]):
     def __init__(self, pattern: AnyStr, options: Options) -> None: ...
-    @overload
-    def search(
-        self: _Regexp[str], text: str, pos: int | None = None, endpos: int | None = None
-    ) -> _Match[str] | None: ...
-    @overload
+
     def search(
-        self: _Regexp[bytes],
-        text: bytes,
-        pos: int | None = None,
-        endpos: int | None = None,
-    ) -> _Match[bytes] | None: ...
-    @overload
-    def match(
-        self: _Regexp[str], text: str, pos: int | None = None, endpos: int | None = None
-    ) -> _Match[str] | None: ...
-    @overload
+        self, text: AnyStr, pos: int | None = None, endpos: int | None = None
+    ) -> _Match[AnyStr] | None: ...
+
     def match(
-        self: _Regexp[bytes],
-        text: bytes,
-        pos: int | None = None,
-        endpos: int | None = None,
-    ) -> _Match[bytes] | None: ...
-    @overload
-    def fullmatch(
-        self: _Regexp[str], text: str, pos: int | None = None, endpos: int | None = None
-    ) -> _Match[str] | None: ...
-    @overload
+        self, text: AnyStr, pos: int | None = None, endpos: int | None = None
```

Comment:
Context: The point of `bytes` vs `AnyStr` in the stubs is to convey the fact that if the input is a `str` the output will be a `Match[str]`. If we used `AnyStr`, then input were `str` then the output could be either `Match[bytes]` or `Match[str]` which loses some typing constraints.

Anyway, happy to just take the missing overload without the AnyStr change if you could scope down this PR to just that.

Reply:
thanks, looks like I misunderstood `AnyStr` here

Reply:
then there is a error in typing for `sub` I think.

this is valid at runtime, 

```python
from __future__ import annotations

import re2

pattern: re2._Regexp[str] = re2.compile(r"1")

def repl(m: re2._Match[bytes]) -> bytes:
    return b"a"

result: bytes = pattern.sub(repl, b" 0 1 2 ")
print(result)
```

which means `def sub(self: _Regexp[str], repl: str, text: str, count: int = 0) -> str` should just be `def sub(self, repl: str, text: str, count: int = 0) -> str`

Reply:
also this is valid `re2.findall(r"1", b"0 1 2"))`, and it doesn't match one of these:

```
@overload
def findall(
    pattern: _Regexp[str] | str, text: str, options: Options | None = None
) -> list[str]: ...
@overload
def findall(
    pattern: _Regexp[bytes] | bytes, text: bytes, options: Options | None = None
) -> list[bytes]: ...
```

Reply:
Thanks for the investigation!

I based the stubs on the idea that `re2` would be used as a drop-in for `re` so I made the constraints stronger than `re2` itself supports, but I don't have an issue with making things match the `re2` implementation more.

The main issue with that though is since this repo isn't directly related to the `re2` one it is easy for the stubs to drift from the implementation in a way that wouldn't be helpful for stub users, but we can discuss that on the next PR if you want.

Reply:
Looking at the in progress stubs produced by the re2 maintainers, https://github.com/google/re2/issues/496, it looks like they are more in line with your original `AnyStr` change.

Reply:
yes, which mean using typing from `re` is fine.

Maybe we can just lock version to `google-re2==1.1.*` so downstream users will fail to resolve deps if re2 release 1.2, and we can check typing then.

Reply:
> Looking at the in progress stubs produced by the re2 maintainers, [google/re2#496](https://github.com/google/re2/issues/496), it looks like they are more in line with your original `AnyStr` change.

can you point out where the file is? I didn't find it in that issue

Reply:
OK, I find it at https://github.com/google/re2/compare/main...cosmicexplorer:type-stubs?expand=1

But it's in-complete, as I explained before.